Completion on ":set filetype="

arp242 opened the pull request:
":filetype <Tab>" and ":set filetype=<Tab>" will now complete known
filetypes.

There's a very old patch for this in #285. I didn't add the "set
syntax=" completion from that, since it uses the list of filetypes
rather than syntax files (which is wrong).

arp242 committed:
Completion ":set filetype="

":set filetype=<Tab>" will now complete known filetypes.

There's a very old patch for this in #285. I didn't add the "set
syntax=" completion from that, since it uses the list of filetypes
rather than syntax files (which is wrong).

brammool commented:
Please add tests.  For the set command in test_options.vim  Test_set_completion().
For the :filetype command the argument is not a filetype name but "on", "off", "plugin on", etc.

lacygoill commented:
This adds a test for the completion after `set filetype=`:
```diff
diff --git a/src/testdir/test_options.vim b/src/testdir/test_options.vim
index cc02c5c7c..f161a8fb3 100644
--- a/src/testdir/test_options.vim
+++ b/src/testdir/test_options.vim
@@ -332,6 +332,10 @@ func Test_set_completion()
   call feedkeys(":set key=\<Tab>\<C-B>\"\<CR>", 'xt')
   call assert_equal('"set key=*****', @:)
   set key=
+
+  " Expand filetypes for 'filetype'
+  call feedkeys(":set filetype=a\<C-A>\<C-B>\"\<CR>", 'xt')
+  call assert_equal("\"set filetype=" .. getcompletion('a*', 'filetype')->join(), @:)
 endfunc
 
 func Test_set_errors()
```
But it doesn't fix the expansion of `:filetype` which still wrongly suggests filetypes while it should suggest the keywords listed at [`:h :filetype-overwiew`](https://vimhelp.org/filetype.txt.html#:filetype-overview):

    :filetype on                    on              unchanged       unchanged
    :filetype off                   off             unchanged       unchanged
    :filetype plugin on             on              on              unchanged
    :filetype plugin off            unchanged       off             unchanged
    :filetype indent on             on              unchanged       on
    :filetype indent off            unchanged       unchanged       off
    :filetype plugin indent on      on              on              on
    :filetype plugin indent off     unchanged       off             off

lacygoill commented:
I've tried to finish the PR.  This is what I have so far:
```diff
diff --git a/src/cmdexpand.c b/src/cmdexpand.c
index d51f5c642..ad162a235 100644
--- a/src/cmdexpand.c
+++ b/src/cmdexpand.c
@@ -1747,6 +1747,10 @@ set_one_cmd_context(
 	    set_context_in_profile_cmd(xp, arg);
 	    break;
 #endif
+	case CMD_filetype:
+	    set_context_in_filetype_cmd(xp, arg);
+	    break;
+
 	case CMD_behave:
 	    xp->xp_context = EXPAND_BEHAVE;
 	    xp->xp_pattern = arg;
@@ -2141,6 +2145,7 @@ ExpandFromContext(
 # ifdef FEAT_PROFILE
 	    {EXPAND_PROFILE, get_profile_name, TRUE, TRUE},
 # endif
+	    {EXPAND_FILETYPECMD, get_filetype_arg, TRUE, TRUE},
 # if defined(HAVE_LOCALE_H) || defined(X_LOCALE)
 	    {EXPAND_LANGUAGE, get_lang_arg, TRUE, FALSE},
 	    {EXPAND_LOCALES, get_locales, TRUE, FALSE},
diff --git a/src/ex_docmd.c b/src/ex_docmd.c
index 8b9db6812..920a39f4b 100644
--- a/src/ex_docmd.c
+++ b/src/ex_docmd.c
@@ -9089,6 +9089,69 @@ ex_filetype(exarg_T *eap)
 	semsg(_(e_invarg2), arg);
 }
 
+// Command line expansion for :filetype.
+static enum
+{
+    FEXP_SUBCMD,	// expand :filetype sub-commands
+} fexpand_what;
+
+static char *fexpand_cmds[] = {
+                        "on",
+#define FILETYPECMD_ON		0
+                        "off",
+#define FILETYPECMD_OFF		1
+                        "plugin",
+#define FILETYPECMD_PLUGIN	2
+                        "indent",
+#define FILETYPECMD_INDENT	3
+                        NULL
+#define FILETYPECMD_LAST	4
+};
+
+/*
+ * Function given to ExpandGeneric() to obtain the filetype command
+ * specific expansion.
+ */
+    char_u *
+get_filetype_arg(expand_T *xp UNUSED, int idx)
+{
+    switch (fexpand_what)
+    {
+    case FEXP_SUBCMD:
+        return (char_u *)fexpand_cmds[idx];
+    default:
+        return NULL;
+    }
+}
+
+/*
+ * Handle command line completion for :filetype command.
+ */
+    void
+set_context_in_filetype_cmd(expand_T *xp, char_u *arg)
+{
+    char_u	*end_subcmd;
+
+    // Default: expand subcommands.
+    xp->xp_context = EXPAND_FILETYPECMD;
+    fexpand_what = FEXP_SUBCMD;
+    xp->xp_pattern = arg;
+
+    end_subcmd = skiptowhite(arg);
+    if (*end_subcmd == NUL)
+        return;
+
+    if (end_subcmd - arg == 6 && (STRNCMP(arg, "indent", 6) == 0
+                               || STRNCMP(arg, "plugin", 6) == 0))
+    {
+        xp->xp_context = EXPAND_FILETYPECMD;
+        xp->xp_pattern = skipwhite(end_subcmd);
+        return;
+    }
+
+    xp->xp_context = EXPAND_NOTHING;
+}
+
 /*
  * ":setfiletype [FALLBACK] {name}"
  */
diff --git a/src/option.c b/src/option.c
index b4893a10a..7a79d2e97 100644
--- a/src/option.c
+++ b/src/option.c
@@ -6209,6 +6209,12 @@ set_context_in_set_cmd(
 	    else
 		xp->xp_backslash = XP_BS_ONE;
 	}
+	else if (p == (char_u *)&p_ft) {
+	    xp->xp_context = EXPAND_FILETYPE;
+	}
+	else if (p == (char_u *)&p_ft) {
+	    xp->xp_context = EXPAND_FILETYPE;
+	}
 	else
 	{
 	    xp->xp_context = EXPAND_FILES;
diff --git a/src/optiondefs.h b/src/optiondefs.h
index 6cea0177c..9961bd477 100644
--- a/src/optiondefs.h
+++ b/src/optiondefs.h
@@ -946,7 +946,7 @@ static struct vimoption options[] =
 				    (char_u *)FALSE,
 #endif
 					(char_u *)0L} SCTX_INIT},
-    {"filetype",    "ft",   P_STRING|P_ALLOCED|P_VI_DEF|P_NOGLOB|P_NFNAME,
+    {"filetype",    "ft",   P_STRING|P_EXPAND|P_ALLOCED|P_VI_DEF|P_NOGLOB|P_NFNAME,
 			    (char_u *)&p_ft, PV_FT,
 			    {(char_u *)"", (char_u *)0L}
 			    SCTX_INIT},
diff --git a/src/proto/ex_cmds.pro b/src/proto/ex_cmds.pro
index 9036dd205..956b58d31 100644
--- a/src/proto/ex_cmds.pro
+++ b/src/proto/ex_cmds.pro
@@ -39,4 +39,5 @@ void ex_smile(exarg_T *eap);
 void ex_drop(exarg_T *eap);
 char_u *skip_vimgrep_pat(char_u *p, char_u **s, int *flags);
 void ex_oldfiles(exarg_T *eap);
+void set_context_in_filetype_cmd(expand_T *xp, char_u *arg);
 /* vim: set ft=c : */
diff --git a/src/proto/ex_docmd.pro b/src/proto/ex_docmd.pro
index 94770f2e3..d55de7052 100644
--- a/src/proto/ex_docmd.pro
+++ b/src/proto/ex_docmd.pro
@@ -61,4 +61,5 @@ void set_no_hlsearch(int flag);
 int is_loclist_cmd(int cmdidx);
 int get_pressedreturn(void);
 void set_pressedreturn(int val);
+char_u *get_filetype_arg(expand_T *xp, int idx);
 /* vim: set ft=c : */
diff --git a/src/testdir/test_options.vim b/src/testdir/test_options.vim
index c8b2700dd..19c62cc7b 100644
--- a/src/testdir/test_options.vim
+++ b/src/testdir/test_options.vim
@@ -332,6 +332,14 @@ func Test_set_completion()
   call feedkeys(":set key=\<Tab>\<C-B>\"\<CR>", 'xt')
   call assert_equal('"set key=*****', @:)
   set key=
+
+  " Expand filetypes for 'filetype'
+  call feedkeys(":set filetype=a\<C-A>\<C-B>\"\<CR>", 'xt')
+  call assert_equal('"set filetype=' .. getcompletion('a*', 'filetype')->join(), @:)
+
+  " Expand :filetype arguments
+  call feedkeys(":filetype \<C-A>\<C-B>\"\<CR>", 'xt')
+  call assert_equal('"filetype indent off on plugin', @:)
 endfunc
 
 func Test_set_errors()
diff --git a/src/vim.h b/src/vim.h
index de2482e8b..2d147af27 100644
--- a/src/vim.h
+++ b/src/vim.h
@@ -777,6 +777,7 @@ extern int (*dyn_libintl_wputenv)(const wchar_t *envstring);
 #define EXPAND_MAPCLEAR		47
 #define EXPAND_ARGLIST		48
 #define EXPAND_DIFF_BUFFERS	49
+#define EXPAND_FILETYPECMD	50
 
 // Values for exmode_active (0 is no exmode)
 #define EXMODE_NORMAL		1
```
It includes tests and fixes the wrong completion for `:filetype`.

But I think it needs more work, as I don't know C.  Some variables should probably be renamed, and I'm not sure about where to put the new code relative to the existing one.

---

Also, after `:filetype`, if we've just completed `plugin`, the next completion should not suggest `plugin` again.  Same thing with `indent`.

Besides, the patch lets us complete up to 2 arguments; we should be able to complete up to 3 arguments.

Finally, the suggestions should never include something which generates an invalid command.  For example, after `:filetype plugin indent`, and `filetype indent plugin` only `on` and `off` should be suggested.

Not sure how to fix these issues.  I copied some code from the `:profile` completion, which is close enough to `:filetype`, but not enough for me to fix those issues.  If anyone can think of another command which has a completion more similar to `:filetype`, maybe I could look at how it's implemented and use whatever is relevant.

Alternatively, if someone else wants to finish the patch, [here are my notes](https://github.com/lacygoill/wiki/blob/master/vim/patch.md).

---

Something else is weird.  We can't complete `set ft=`; we need to type at least 1 character after the equal sign for the completion to kick in.  There's no such requirement with `:setf`.  Not sure why the difference.  It doesn't come from my patch.  The issue was already present in the original PR.

arp242 commented:
> For the :filetype command the argument is not a filetype name but "on", "off", "plugin on", etc.

I got confused with `:setfiletype` command (which already has completion); how silly of me. I removed this part, and added a test.

> Something else is weird. We can't complete set ft=; we need to type at least 1 character after the equal sign for the completion to kick in. There's no such requirement with :setf. Not sure why the difference. It doesn't come from my patch. The issue was already present in the original PR.

This is how completions for all the options work at the moment: it will first complete the current value (which may be blank). You can also see this with `path` for example:

    :set path=
    :set path=<Tab>      Nothing happens
    :set path=/<Tab>     Completion!

I agree this is really confusing behaviour, but also not really something to fix/change in this particular patch. I intended to look a bit at this after I finished this to see if I could find the rationale for doing it like this, and if it could be improved.

Re: your test; I've added that, thanks! I wrote pretty much the same myself the other day, but I ran in to some problems with the test runner and it took me a while that it's running against the system-installed Vim instead of the changes I made, do'h >_<

arp242 commented:
I created a new issue for the completion: #7769; and I'll look at adding proper support for `:filetype` (as well as other commands that don't have completion, this has been an annoyance for some time) in a new PR. The only reason I combined them in a single patch is because I got confused between `:filetype` and `:setfiletype` 🤦

Anyway, I think this should be grand to merge now @brammool

lacygoill commented:
> as well as other commands that don't have completion, this has been an annoyance for some time

The ones in which I would be the most interested in are `:breakadd` and `:profile`.  Those are usually executed when something goes wrong.  In those cases, having something else also going wrong (like an expansion which fails to suggest anything) is really annoying.

Relevant todo item:

https://github.com/vim/vim/blob/e7bebc495d4014d7bc81f863939c35268cb8e97d/runtime/doc/todo.txt#L437-L440

Quoted for people reading from the mailing list:

   > Implement completion for "breakadd".  Should expand the second argument, e.g.
   > "func", and then function names after ":breakadd func".  Including
   > script-local functions.
   > Also for ":profile".
